### PR TITLE
Refactor ExpressionRunner

### PR DESCRIPTION
**Pull request by dcodeIO**
Tackles the concerns raised in https://github.com/WebAssembly/binaryen/issues/2797 directly related to https://github.com/WebAssembly/binaryen/pull/2702 by reverting merging all of `PrecomputeExpressionRunner` into the base `ExpressionRunner`, instead adding a common base for both the precompute pass and the new C-API to inherit. No functional changes.

---

### Current hierarchy after https://github.com/WebAssembly/binaryen/pull/2702 is

```
ExpressionRunner
├ [PrecomputeExpressionRunner]
├ [CExpressionRunner]
├ ConstantExpressionRunner
└ RuntimeExpressionRunner
```

where `ExpressionRunner` contains functionality not utilized by `ConstantExpressionRunner` and `RuntimeExpressionRunner`.

### New hierarchy will be:

```
ExpressionRunner
├ ConstantExpressionRunner
│  ├ [PrecomputeExpressionRunner]
│  └ [CExpressionRunner]
├ InitializerExpressionRunner
└ RuntimeExpressionRunner
```

with the precompute pass's and the C-API's shared functionality now moved out of `ExpressionRunner` into a new `ConstantExpressionRunner`. Also renames the previous `ConstantExpressionRunner` to `InitializerExpressionRunner` to [better represent its uses](https://webassembly.org/docs/modules/#initializer-expression) and to make its previous name usable for the new intermediate template, where it fits perfectly. Also adds a few comments answering some of the questions that came up recently.

### Old hierarchy before https://github.com/WebAssembly/binaryen/pull/2702 for comparison:

```
ExpressionRunner
├ [PrecomputeExpressionRunner]
├ ConstantExpressionRunner
└ RuntimeExpressionRunner
```

cc @aheejin

### CODE REVIEW

**Comment by dcodeIO** [comment]
Also note that `InitializerExpressionRunner` in its current form seems problematic as it might support more expressions than it should (or even abort on invalid initializers instead of emitting a validation error?). This is unrelated to this PR and #2702 however and used to be this way already before, so should probably be covered in a separate PR.

File: src/wasm-interpreter.h [hunk 1]
```diff
@@ -149,52 +149,13 @@ class Indenter {
...
-  // we are not interested in any context or if the subclass uses an alternative
-  // mechanism, like RuntimeExpressionRunner does.
-  Module* module = nullptr;
-
-  // Flags indicating special requirements. See FlagValues.
-  Flags flags = FlagValues::DEFAULT;
-
   // Maximum depth before giving up.
-  Index maxDepth = NO_LIMIT;
+  Index maxDepth;
```
Comment (dcodeIO) [inline src/wasm-interpreter.h:154]: In https://github.com/WebAssembly/binaryen/pull/2702 it has been requested to initialize both ´maxDepth` and `maxLoopIterations` here with `= NO_LIMIT``. However, the single constructor now sets these while allowing to omit either just `maxLoopIterations` or both. So, C++ newbie question: Are there scenarios where a derived template may never call that constructor?

**Comment by dcodeIO** [comment]
Also, I broke CI again. But this time I'm not able to fix it apparently :(

**Comment by dcodeIO** [comment]
Also, I fixed CI again!

**Comment by aheejin** [comment]
Thank you for the quick refactoring! Sorry I still may not have the full context yet, but why are `ConstantExpressionRunner` and `PrecomputeExpressionRunner` two separate classes? It seems the functionalities you put into `ExpressionRunner` (and moved to `ConstantExpressionRunner` here) are mostly about precomputing, so what I was wondering is why we can't just improve and expose the existing `PrecomputeExpressionRunner` and use it in the Binaryen and C API in the first place. (This will require moving `PrecomputeExpressionRunner` out of Precompute.cpp, but I don't think this will be a problem)

**Comment by dcodeIO** [comment]
> why are ConstantExpressionRunner and PrecomputeExpressionRunner two separate classes?

The precompute pass has a special mechanism that the C-API does not have, in that it computes a map of `GetValues` (known values of `local.get`s by reference, not by index) during the pass and [uses these as additional context](https://github.com/WebAssembly/binaryen/pull/2804/files#diff-633311eb1c9f14a0358c77ab7ed93b97R80-R86) of the runner. The base classes has similar code, but by index. An earlier attempt was to just make this an optional feature of what's now `ConstantExpressionRunner`, but this plus the fact that the C-API's runner must be `final` so we can `delete` its instances contributed to the general confusion during the earlier PR I believe.

>  It seems the functionalities you put into ExpressionRunner (and moved to ConstantExpressionRunner here) are mostly about precomputing

Yep, one way to look at this is that `ConstantExpressionRunner` is the new exposed `PrecomputeExpressionRunner`, with the original `PrecomputeExpressionRunner` now only adding the `GetValues` mechanism that is irrelevant outside of the precompute pass.

> I was wondering is why we can't just improve and expose the existing PrecomputeExpressionRunner and use it in the Binaryen and C API in the first place

So, yeah, that's kinda what this is now after the refactor. But I think `Constant` is a slightly better name, since it really "Executes a suspected constant expression", while the `Precompute` (replacement of expressions) happens elsewhere.

Does that make sense? :)

**Comment by aheejin** [comment]
> The precompute pass has a special mechanism that the C-API does not have, in that it computes a map of `GetValues` (known values of `local.get`s by reference, not by index) during the pass and [uses these as additional context](https://github.com/WebAssembly/binaryen/blob/35b60af10f08f560268a512edbfdbe8d09e21093/src/passes/Precompute.cpp#L80-L86) of the runner. The base class has similar code, but by index.

How are the two mechanisms (`GetValues`-based vs. index-based) different? Do we need both?

> An earlier attempt was to just make this an optional feature of what's now `ConstantExpressionRunner`, but this plus the fact that the C-API's runner must be `final` so we can `delete` its instances contributed to the general confusion during the earlier PR I believe.

How is a class being `final` relevant to whether we can delete it or not?

**Comment by tlively** [comment]
See https://github.com/WebAssembly/binaryen/pull/2702#discussion_r395162786 for the context of the `final` thing.

**Comment by dcodeIO** [comment]
> How are the two mechanisms (GetValues-based vs. index-based) different? Do we need both?

Iiuc the GetValues-based mechanism is specialized to work with LocalGraph after its work is done, while the index-based approach observes in execution order. Would assume that there are ways to unify both, somehow. Currently, precompute uses GetValues to establish an initial link of gets to values, and will then resort to the index-based approach on new sets and their respective gets during linear execution of the expression of interest. The challenge I faced in my PR was that one cannot easily compute the indexed variant from GetValues, since execution order is lost when that map is created, so one doesn't know which of all the values to initialize the runner with.

**Comment by dcodeIO** [comment]
> It looks CExpressionRunner is there only to be marked as final, right? Can we delete it and use ConstantExpressionRunner from C API instead?

Sure, let me see. How would you model this, considering that CExpressionRunner is a class, and ConstantExpressionRunner is its template? My intuition would be to make ConstantExpressionRunner a class, with a `virtual visitLocalGet` for `PrecomputingExpressionRunner` to overload, and to move `ConstantExpressionRunner`'s implementation to `src/wasm/wasm-interpreter.cpp`?

**Comment by aheejin** [comment]
Hmm, yeah, come to think of it, it's not gonna be simple :( Making those methods virtual will break the static polymorphism this class hierarchy is using. I think the problem here stems from the fact that even though `ConstantExpressionRunner` is a parent class of `PrecoputeExpressionRunner`, their `visitLocalGet` and `visitLocalSet` are actually two different mechanisms.

How about this hierarchy, where `ConstantExpressionRunner` does not have local-related methods itself, and make a new class `LocalExpressionRunner` (I'm not sure about the name so I just threw some random name) and it will have index-based local handling methods. So `ConstantExpressionRunner` has common methods used between `PrecomputeExpressionRunner` and `LocalExpressionRunner`, and `PrecomputeExpressionRunner` and `LocalExpressionRunner` have their respective local-related handling methods. (`PrecomputeExpressionRunner` has `GetValues`-based methods and `LocalExpressionRunner` has index-based methods)
```
ExpressionRunner
├ ConstantExpressionRunner
│  ├ [PrecomputeExpressionRunner]
│  └ [LocalExpressionRunner]
├ InitializerExpressionRunner
└ RuntimeExpressionRunner
```

Do you think this makes sense? Please let me know if you have better ideas. Thanks for your time!

**Comment by dcodeIO** [comment]
> Making those methods virtual will break the static polymorphism this class hierarchy is using

There's already a `virtual trap()` method inherited from `ExpressionRunner`, but I guess this is mostly about avoiding virtual methods where we can, that correct?

> and PrecomputeExpressionRunner has GetValues-based methods and LocalExpressionRunner has index-based methods

Hmm, might work. So far I assumed that `PrecomputeExpressionRunner` still benefits from being able to fall back to the index-based variant, but might be that `GetValues` already covers everything and that's not necessary? If that's the case, perhaps there's indeed a way to convert initial `setLocalValue` values to `GetValues` by spawning `LocalGraph` prior to executing the runner, also solving https://github.com/WebAssembly/binaryen/pull/2787.

**Comment by aheejin** [comment]
> There's already a `virtual trap()` method inherited from `ExpressionRunner`, but I guess this is mostly about avoiding virtual methods where we can, that correct?

Yes. `OverriddenVisitor` and its children are expected to override `visit***` methods and they are designed to be [statically called](https://github.com/WebAssembly/binaryen/blob/12445969dc5b32cbe82b829fe8c061f5d98fcbe7/src/wasm-traversal.h#L285-L287) to avoid vtable traversal.

> Hmm, might work. So far I assumed that `PrecomputeExpressionRunner` still benefits from being able to fall back to the index-based variant, but might be that `GetValues` already covers everything and that's not necessary? If that's the case, perhaps there's indeed a way to convert initial `setLocalValue` values to `GetValues` by spawning `LocalGraph` prior to executing the runner (for `CExpressionRunner` only), also solving #2787.

I'm not sure what this paragraph means. I'm not suggesting to add any new functionality or `LocalGraph` or anything. What I meant was I was not sure why `ConstantExpressionRunner` has to be a parent class of `PrecomputeExpressionRunner`. Are there any routines that `PrecomputeExpressionRunner` inherit from `ConstantExpressionRunner`? If there is no common functionality between `PrecomputeExpressionRunner` and `ConstantExpressionRunner`, is this simpler hierarchy possible?
```
ExpressionRunner
├ PrecomputeExpressionRunner (`GetValues`-based local methods)
├ LocalExpressionRunner (index-based local methods + call traversal)
├ InitializerExpressionRunner
└ RuntimeExpressionRunner
```

**Comment by dcodeIO** [comment]
> Are there any routines that PrecomputeExpressionRunner inherit from ConstantExpressionRunner

It [inherits practically everything](https://github.com/WebAssembly/binaryen/blob/522a990cb2a4b06e559bf50ba8238982b1c5c383/src/passes/Precompute.cpp#L48-L90) with this PR, except that it overrides `visitLocalGet` to inject `GetValues` from the precompute pass:

```
ExpressionRunner                    Stub with lots of "unimp"
├ ConstantExpressionRunner          Huge, common functionality of Precompute/C-API
│  ├ [PrecomputeExpressionRunner]   Tiny (ctor + visitLocalGet)
│  └ [CExpressionRunner]            Tiny (ctor)
├ InitializerExpressionRunner       Something completely different
└ RuntimeExpressionRunner           Something completely different
```

To me that looks quite reasonable already, and I have been under the impression that we are now trying to unify PrecomputeExpressionRunner/CExpressionRunner even more, but perhaps this is a misunderstanding? Like, PrecomputeExpressionRunner doesn't actually duplicate any code, which I believe you mentioned earlier?

**Comment by kripken** [comment]
Looks very nice!

**Pull request merged**